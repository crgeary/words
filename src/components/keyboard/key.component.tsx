import classNames from "classnames";
import { ComponentPropsWithoutRef, FC } from "react";
import { CharState } from "../../types/char-state";

export type KeyProps = ComponentPropsWithoutRef<"button"> & {
    state?: CharState;
};

export const Key: FC<KeyProps> = ({ children, className, state = CharState.Idle, ...props }) => {
    return (
        <button
            className={classNames(
                "inline-flex justify-center items-center px-1 py-3 border shadow-sm rounded tracking-tight uppercase font-medium text-sm text-center active:shadow-inner",
                className,
                {
                    "bg-green-500 border-green-600 text-green-50": CharState.Correct === state,
                    "bg-yellow-500 border-yellow-600 text-yellow-900": CharState.Present === state,
                    "bg-slate-100 border-slate-200 text-slate-300": CharState.Absent === state,
                    "bg-white border-violet-200 text-slate-600": CharState.Idle === state,
                },
            )}
            {...props}
        >
            {children}
        </button>
    );
};
