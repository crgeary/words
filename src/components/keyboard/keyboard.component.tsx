import { BackspaceIcon } from "@heroicons/react/outline";
import { FC, useEffect } from "react";
import { getKeyboardState } from "../../utils/get-keyboard-state";
import { Key } from "./key.component";

export enum KeyCodes {
    Enter = "Enter",
    Backspace = "Backspace",
}

export type KeyboardProps = {
    solution: string;
    guesses: string[];
    onChar: (char: string) => void;
    onEnter: () => void;
    onDelete: () => void;
};

export const Keyboard: FC<KeyboardProps> = ({ solution, guesses, onChar, onEnter, onDelete }) => {
    const states = getKeyboardState(guesses, solution);

    useEffect(() => {
        const handler = ({ key }: KeyboardEvent) => {
            if (key === KeyCodes.Enter) {
                onEnter();
            } else if (key === KeyCodes.Backspace) {
                onDelete();
            } else {
                if (key.length === 1 && key.toLowerCase() >= "a" && key.toUpperCase() <= "z") {
                    onChar(key);
                }
            }
        };
        window.addEventListener("keyup", handler);
        return () => {
            window.removeEventListener("keyup", handler);
        };
    });

    return (
        <div className="flex flex-col gap-1">
            <div className="flex justify-center -mx-0.5">
                {`qwertyuiop`.split("").map((key) => (
                    <Key
                        key={key}
                        state={states[key]}
                        className="flex-grow flex-shrink basis-0 mx-0.5"
                        onTouchStart={() => window?.navigator?.vibrate?.(75)}
                        onClick={() => onChar(key)}
                        data-testid={`key-${key}`}
                    >
                        {key}
                    </Key>
                ))}
            </div>
            <div className="flex justify-center -mx-0.5">
                <span className="flex-grow-0.5 flex-shrink basis-0"></span>
                {`asdfghjkl`.split("").map((key) => (
                    <Key
                        key={key}
                        state={states[key]}
                        className="flex-grow flex-shrink basis-0 mx-0.5"
                        onTouchStart={() => window?.navigator?.vibrate?.(75)}
                        onClick={() => onChar(key)}
                        data-testid={`key-${key}`}
                    >
                        {key}
                    </Key>
                ))}
                <span className="flex-grow-0.5 flex-shrink basis-0"></span>
            </div>
            <div className="flex justify-center -mx-0.5">
                <Key
                    className="flex-grow-1.5 flex-shrink basis-0 mx-0.5"
                    onTouchStart={() => window?.navigator?.vibrate?.(75)}
                    onClick={() => onEnter()}
                    data-testid="key-enter"
                >
                    <span className="text-xs">Enter</span>
                    <span className="sr-only">Enter</span>
                </Key>
                {`zxcvbnm`.split("").map((key) => (
                    <Key
                        key={key}
                        state={states[key]}
                        className="flex-grow flex-shrink basis-0 mx-0.5"
                        onTouchStart={() => window?.navigator?.vibrate?.(75)}
                        onClick={() => onChar(key)}
                        data-testid={`key-${key}`}
                    >
                        {key}
                    </Key>
                ))}
                <Key
                    className="flex-grow-1.5 flex-shrink basis-0 mx-0.5"
                    onTouchStart={() => window?.navigator?.vibrate?.(75)}
                    onClick={() => onDelete()}
                    data-testid="key-delete"
                >
                    <BackspaceIcon className="w-5 h-5" />
                    <span className="sr-only">Delete</span>
                </Key>
            </div>
        </div>
    );
};
