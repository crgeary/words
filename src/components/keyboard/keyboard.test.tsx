import { render } from "@testing-library/react";
import { Keyboard } from "./keyboard.component";

test("it should match snapshot", () => {
    const { container } = render(
        <Keyboard
            solution="heros"
            guesses={["hello", "world"]}
            onChar={() => {}}
            onEnter={() => {}}
            onDelete={() => {}}
        />,
    );
    expect(container.firstChild).toMatchSnapshot();
});
