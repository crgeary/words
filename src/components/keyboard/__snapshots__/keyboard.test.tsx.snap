// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should match snapshot 1`] = `
<div
  class="flex flex-col gap-1"
>
  <div
    class="flex justify-center -mx-0.5"
  >
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-q"
    >
      <span
        class="uppercase"
      >
        q
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-slate-100 border-slate-200 text-slate-300"
      data-testid="key-w"
    >
      <span
        class="uppercase"
      >
        w
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-green-500 border-green-600 text-green-50"
      data-testid="key-e"
    >
      <span
        class="uppercase"
      >
        e
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-green-500 border-green-600 text-green-50"
      data-testid="key-r"
    >
      <span
        class="uppercase"
      >
        r
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-t"
    >
      <span
        class="uppercase"
      >
        t
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-y"
    >
      <span
        class="uppercase"
      >
        y
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-u"
    >
      <span
        class="uppercase"
      >
        u
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-i"
    >
      <span
        class="uppercase"
      >
        i
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-yellow-500 border-yellow-600 text-yellow-9000"
      data-testid="key-o"
    >
      <span
        class="uppercase"
      >
        o
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-p"
    >
      <span
        class="uppercase"
      >
        p
      </span>
    </button>
  </div>
  <div
    class="flex justify-center -mx-0.5"
  >
    <span
      class="flex-grow-0.5 flex-shrink basis-0"
    />
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-a"
    >
      <span
        class="uppercase"
      >
        a
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-s"
    >
      <span
        class="uppercase"
      >
        s
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-slate-100 border-slate-200 text-slate-300"
      data-testid="key-d"
    >
      <span
        class="uppercase"
      >
        d
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-f"
    >
      <span
        class="uppercase"
      >
        f
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-g"
    >
      <span
        class="uppercase"
      >
        g
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-green-500 border-green-600 text-green-50"
      data-testid="key-h"
    >
      <span
        class="uppercase"
      >
        h
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-j"
    >
      <span
        class="uppercase"
      >
        j
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-k"
    >
      <span
        class="uppercase"
      >
        k
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-slate-100 border-slate-200 text-slate-300"
      data-testid="key-l"
    >
      <span
        class="uppercase"
      >
        l
      </span>
    </button>
    <span
      class="flex-grow-0.5 flex-shrink basis-0"
    />
  </div>
  <div
    class="flex justify-center -mx-0.5"
  >
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow-1.5 flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-enter"
    >
      <span
        class="uppercase"
      >
        Enter
        <span
          class="sr-only"
        >
          Enter
        </span>
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-z"
    >
      <span
        class="uppercase"
      >
        z
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-x"
    >
      <span
        class="uppercase"
      >
        x
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-c"
    >
      <span
        class="uppercase"
      >
        c
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-v"
    >
      <span
        class="uppercase"
      >
        v
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-b"
    >
      <span
        class="uppercase"
      >
        b
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-n"
    >
      <span
        class="uppercase"
      >
        n
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-m"
    >
      <span
        class="uppercase"
      >
        m
      </span>
    </button>
    <button
      class="inline-flex justify-center items-center p-2 border shadow-sm rounded font-medium text-sm text-center active:shadow-inner flex-grow-1.5 flex-shrink basis-0 mx-0.5 bg-white border-violet-200 text-slate-600"
      data-testid="key-delete"
    >
      <span
        class="uppercase"
      >
        <svg
          aria-hidden="true"
          class="w-5 h-5"
          fill="none"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 14l2-2m0 0l2-2m-2 2l-2-2m2 2l2 2M3 12l6.414 6.414a2 2 0 001.414.586H19a2 2 0 002-2V7a2 2 0 00-2-2h-8.172a2 2 0 00-1.414.586L3 12z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        <span
          class="sr-only"
        >
          Delete
        </span>
      </span>
    </button>
  </div>
</div>
`;
